core/executor/src/wasmtime/runtime.rs

```diff
@@ -0,0 +1,385 @@
+	let mut context = Context::with_isa(isa, CompilationStrategy::Cranelift);
+
+	// Enable/disable producing of debug info.
+	context.set_debug_info(false);
+
+	// Instantiate and link the env module.
+	let global_exports = context.get_global_exports();
+	let env_module = instantiate_env_module(global_exports)?;
+	context.name_instance("env".to_owned(), env_module);
+
+	// Compile the wasm module.
+	let module = context.compile_module(&code)
+		.map_err(WasmError::WasmtimeSetup)?;
+
+	Ok((module, context))
+}
+
+/// Call a function inside a precompiled Wasm module.
+fn call_method(
+	context: &mut Context,
+	module: &mut CompiledModule,
+	ext: &mut dyn Externalities,
+	method: &str,
+	data: &[u8],
+	heap_pages: u32,
+) -> Result<Vec<u8>> {
+	// Old exports get clobbered in `InstanceHandle::new` if we don't explicitly remove them first.
+	//
+	// The global exports mechanism is temporary in Wasmtime and expected to be removed.
+	// https://github.com/CraneStation/wasmtime/issues/332
+	clear_globals(&mut *context.get_global_exports().borrow_mut());
+
+	let mut instance = module.instantiate()
+		.map_err(SetupError::Instantiate)
+		.map_err(ActionError::Setup)
+		.map_err(Error::Wasmtime)?;
+
+	// Ideally there would be a way to set the heap pages during instantiation rather than
+	// growing the memory after the fact. Current this may require an additional mmap and copy.
+	// However, the wasmtime API doesn't support modifying the size of memory on instantiation
+	// at this time.
+	grow_memory(&mut instance, heap_pages)?;
+
+	// Initialize the function executor state.
+	let executor_state = FunctionExecutorState::new(
+		// Unsafely extend the reference lifetime to static. This is necessary because the
+		// host state must be `Any`. This is OK as we will drop this object either before
+		// exiting the function in the happy case or in the worst case on the next runtime
+		// call, which also resets the host state. Thus this reference can never actually
+		// outlive the Context.
+		unsafe { mem::transmute::<_, &'static mut Compiler>(context.compiler_mut()) },
+		get_heap_base(&instance)?,
+	);
+	reset_env_state(context, Some(executor_state))?;
+
+	// Write the input data into guest memory.
+	let (data_ptr, data_len) = inject_input_data(context, &mut instance, data)?;
+	let args = [RuntimeValue::I32(u32::from(data_ptr) as i32), RuntimeValue::I32(data_len as i32)];
+
+	// Invoke the function in the runtime.
+	let outcome = externalities::set_and_run_with_externalities(ext, || {
+		context
+			.invoke(&mut instance, method, &args[..])
+			.map_err(Error::Wasmtime)
+	})?;
+	let trap_error = reset_env_state(context, None)?;
+	let (output_ptr, output_len) = match outcome {
+		ActionOutcome::Returned { values } => {
+			if values.len()	!= 1 {
+				return Err(Error::InvalidReturn);
+			}
+			if let RuntimeValue::I64(val) = values[0] {
+				let output_ptr = <Pointer<u8>>::new(val as u32);
+				let output_len = ((val as u64) >> 32) as usize;
```

Comment:
The fact that we cast it to `u64` first is very important, because right shift on an signed value would give us sign extension (which would be an error). Can we leave a comment on that here?

Reply:
(This should be moved into an helper function that is shared between the wasmi and wasmtime impl)

Reply:
That would be ideal